### PR TITLE
 fix #315 add type apiTokens

### LINKED ISSUE #315: Admin API Key / Secret Gen
1. UI to create API Token
2. UI to delete API Tokens
3. API to power UI (Get, create, delete)

Currently, API authentication is using ADMIN password. Highly sensitive information, we don't want to send passwords around (because it may get compromised).

Instead, we generate API tokens. These tokens can be deleted (hence invalidated) when security of the token is compromised.

All 3rd party (like c0d3.com) will be using API tokens to interact with API.

Keep existing admin password auth. Auth should support both API token and Admin password.

### PR DESCRIPTION
**Pull request by rkalra247**

### CODE REVIEW

File: src/api/apiTokens.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+import express from 'express'
+import uuid4 from 'uuid/v4'
+import { ApiTokens } from '../types/general'
+import { getApiTokens } from '../lib/data'
+
+const apiTokenRouter = express.Router()
+
+apiTokenRouter.post('/', (req, res) => {
+  const allApiTokens = getApiTokens()
+})
```
Comment (hwong0305) [inline src/api/apiTokens.ts]: Add `export default apiTokenRouter`

**Comment by hwong0305** [comment]
Update `src/api/index.ts` to add route for `apiTokenRouter`

File: src/types/general.ts [hunk 2]
```diff
@@ -41,12 +41,18 @@ type ProxyMapping = {
   port?: string
 }
 
+type ApiTokens = {
```
Comment (songz) [inline src/types/general.ts]: 2 issues here:
1. Should be singular. 
2. `AccessToken` is a better name than API token.

File: src/api/apiTokens.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/* eslint @typescript-eslint/camelcase: 0 */
+import express from 'express'
+import uuidv4 from 'uuid/v4'
+import { ApiTokens } from '../types/general'
+import { setData, getApiTokens } from '../lib/data'
+
+const apiTokenRouter = express.Router()
+
+apiTokenRouter.post('/api/tokens', (req, res) => {
```
Comment (songz) [inline src/api/apiTokens.ts]: filename should be `accessTokens`. API route here should simply be:
```
accessTokenRouter.post('/accessTokens', (req, res) => {
  ...
})
```

File: src/types/admin.ts [hunk 4]
```diff
@@ -11,6 +11,7 @@ type DB = {
   serviceKeys: ServiceKey[]
   mappings: Mapping[]
   availableDomains: Domain[]
+  apiTokens: ApiTokens[]
```
Comment (songz) [inline src/types/admin.ts]: accessTokens

File: src/lib/data.ts [hunk 5]
```diff
@@ -1,11 +1,12 @@
...
 import { DB, ServiceKey } from '../types/admin'
-import { Mapping, Domain } from '../types/general'
+import { Mapping, Domain, ApiTokens } from '../types/general'
 
 const data: DB = {
   serviceKeys: [],
   mappings: [],
-  availableDomains: []
+  availableDomains: [],
+  apiTokens: []
```
Comment (songz) [inline src/lib/data.ts]: accessTokens

File: src/lib/data.ts [hunk 6]
```diff
@@ -1,11 +1,12 @@
 import fs from 'fs'
 import { DB, ServiceKey } from '../types/admin'
-import { Mapping, Domain } from '../types/general'
+import { Mapping, Domain, ApiTokens } from '../types/general'
```
Comment (songz) [inline src/lib/data.ts]: AccessToken

File: src/api/accessTokens.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
+/* eslint @typescript-eslint/camelcase: 0 */
```
Comment (songz) [inline src/api/accessTokens.ts]: could you name the file `accessToken` instead of the plural form?

File: src/api/accessTokens.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  setData('apiTokens', allAccessTokens)
+  res.json(tokensObject)
+})
+
+accessTokensRouter.get('/', (req, res) => {
+  res.json([
+    {
+      name: 'hello world'
+    }
+  ])
```
Comment (songz) [inline src/api/accessTokens.ts]: for this, you can just call `getAccessTokens()` and send back all the access tokens.

File: src/tests/integration/accessTokens.test.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import { startAppServer } from '../../server/server'
```
Comment (songz) [inline src/tests/integration/accessTokens.test.ts]: could you rename file to `accessToken.test.js` to keep it plural?

File: src/tests/integration/accessTokens.test.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  beforeAll(async () => {
+    server = await startAppServer(TEST_PORT, ADMIN)
+  })
+
+  afterAll(() => {
+    server.close()
+  })
+
+  it('checks to see if accesstoken is created', async () => {
+    const name = 'c0d3access'
```
Comment (songz) [inline src/tests/integration/accessTokens.test.ts]: you should probably add UUID in here, because we should not be able to create multiple access tokens with the same name. Please also check name in `POST` request to create an accesstoken, make sure you don't create duplicates.

Comment (songz) [reply src/tests/integration/accessTokens.test.ts]: Add another test to check for duplicates names.

File: src/api/accessToken.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+import express from 'express'
+import uuidv4 from 'uuid/v4'
+import { AccessToken } from '../types/general'
+import { setData, getAccessTokens } from '../lib/data'
+
+const accessTokensRouter = express.Router()
+
+accessTokensRouter.post('/', (req, res) => {
+  const allAccessTokens = getAccessTokens()
+  allAccessTokens.find
```
Comment (songz) [inline src/api/accessToken.ts]: remove. this isn't doing anything.

File: src/api/accessToken.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  const tokensObject: AccessToken = {
+    name: req.body.name,
+    id: `${uuidv4()}`
+  }
+  const existingToken = allAccessTokens.find(e => e.name === tokensObject.name)
+  if (existingToken)
+    return res.status(400).json({
+      message: 'This token alreadu exits'
```
Comment (songz) [inline src/api/accessToken.ts]: `already`

File: src/api/accessToken.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
+/* eslint @typescript-eslint/camelcase: 0 */
+import express from 'express'
+import uuidv4 from 'uuid/v4'
+import { AccessToken } from '../types/general'
+import { setData, getAccessTokens } from '../lib/data'
+
+const accessTokensRouter = express.Router()
+
+accessTokensRouter.post('/', (req, res) => {
```
Comment (songz) [inline src/api/accessToken.ts]: could we do a check here? if `req.body.name` length is < 2, then its an invalid name (too short or empty), send back 400.

File: src/tests/integration/accessToken.test.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+  beforeAll(async () => {
+    server = await startAppServer(TEST_PORT, ADMIN)
+  })
+
+  afterAll(() => {
+    server.close()
+  })
+
+  it('checks to see if accesstoken is created', async () => {
```
Comment (songz) [inline src/tests/integration/accessToken.test.ts]: standard way to write test message:
```
it('should create an access token successfully', async() => {.....
```

File: src/tests/integration/accessToken.test.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    const name = `c0d3access${uuidv4()}`
+    const postResponse = await accessTokensAdapter('/', 'POST', {
+      name
+    })
+    const postAccessToken = await postResponse.json()
+    expect(postAccessToken.name).toEqual(name)
```
Comment (songz) [inline src/tests/integration/accessToken.test.ts]: just because we get back a response, does not mean 100% its saved. Could we send a `GET` request here to get all the access tokens, then make sure the one we just created is in the accessToken list?

File: src/tests/integration/accessToken.test.ts [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    const firstName = `c0d3access${uuidv4()}`
+    const postResponse = await accessTokensAdapter('/', 'POST', {
+      name: firstName
+    })
+    expect(postResponse.status).toEqual(200)
+    const secondName = firstName
+    const duplicatePostResponse = await accessTokensAdapter('/', 'POST', {
+      name: secondName
+    })
+    expect(duplicatePostResponse.status).toEqual(400)
```
Comment (songz) [inline src/tests/integration/accessToken.test.ts]: send another `GET` request to get all access tokens, then filter through the array for the name. Filtered array should have length 1.

**Comment by songz** [comment]
rebuild

File: src/tests/integration/accessToken.test.ts [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+  it('should create an access token successfully', async () => {
+    const name = `c0d3access${uuidv4()}`
+    const postResponse = await accessTokensAdapter('/', 'POST', {
+      name
+    })
+    const postAccessToken = await postResponse.json()
+    expect(postAccessToken.name).toEqual(name)
+    const getTokens = await accessTokensAdapter('/', 'GET').then(r => r.json())
+    const tokenExists = getTokens.find(e => e.name === name)
+    if (tokenExists) return `Access token for ${name} was created successfully!`
```
Comment (songz) [inline src/tests/integration/accessToken.test.ts]: Should have a test here instead of return string.
```
const foundToken = getTokens.find(.....)
expect(foundToken.name).toEqual(name)
```
